cmd/diff.go

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright © 2020 Martin Strobel
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package cmd
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"path"
+
+	"github.com/marstr/envelopes"
+	"github.com/marstr/envelopes/persist"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+
+	"github.com/marstr/baronial/internal/index"
+)
+
+type diffCmd struct {
+	cobra.Command
+	Context context.Context
+	Resolver persist.RefSpecResolver
+}
+
+func init() {
+	diff := NewDiffCmd()
+
+	rootCmd.AddCommand(&diff.Command)
+}
+
+func NewDiffCmd() *diffCmd {
+	return NewDiffCmdWithContext(context.Background())
+}
+
+func NewDiffCmdWithContext(ctx context.Context) *diffCmd {
+	retval := &diffCmd{
+		Context: ctx,
+	}
+
+	retval.Command = cobra.Command{
+		Use: "diff [refspec] [refspec]",
+		Short: "Finds the difference between two states, be they from the index or two transactions.",
+		Long: ``,
+		Args: retval.processArgs,
+		PreRunE: setPagedCobraOutput,
+		Run: retval.run,
+	}
+
+	return retval
+}
+
+func (dc diffCmd) processArgs(cmd *cobra.Command, arg []string) error {
+	if err := cobra.MaximumNArgs(2)(cmd, arg); err != nil {
+		return err
+	}
+
+	for _, rs := range arg {
+		id, err := dc.Resolver.Resolve(dc.Context, persist.RefSpec(rs))
+		if err != nil || id.Equal(envelopes.ID{}){
+			return fmt.Errorf("%q is not a valid refspec", rs)
+		}
+	}
+	return nil
+}
+
+func (dc diffCmd) run(cmd *cobra.Command, args []string) {
+	var err error
+	defer func(){
+		if err != nil {
+			logrus.Fatal(err)
```

Comment:
This may prevent the pipe from being closed. All references to `logurus.Fatal` should be removed.